dataclients/kubernetes/ingress.go

```diff
@@ -0,0 +1,775 @@
+	name = name + "_" + id + "_" + strconv.Itoa(index)
+	return routeID(namespace, name, host, "", "")
+}
+
+func setPath(m PathMode, r *eskip.Route, p string) {
+	if p == "" {
+		return
+	}
+
+	switch m {
+	case PathPrefix:
+		r.Predicates = append(r.Predicates, &eskip.Predicate{
+			Name: "PathSubtree",
+			Args: []interface{}{p},
+		})
+	case PathRegexp:
+		r.PathRegexps = []string{p}
+	default:
+		r.PathRegexps = []string{"^" + p}
+	}
+}
+
+func convertPathRule(
+	state *clusterState,
+	metadata *metadata,
+	host string,
+	prule *pathRule,
+	pathMode PathMode,
+) (*eskip.Route, error) {
+
+	ns := metadata.Namespace
+	name := metadata.Name
+
+	if prule.Backend == nil {
+		return nil, fmt.Errorf("invalid path rule, missing backend in: %s/%s/%s", ns, name, host)
+	}
+
+	var (
+		eps []string
+		err error
+		svc *service
+	)
+
+	var hostRegexp []string
+	if host != "" {
+		hostRegexp = []string{"^" + strings.Replace(host, ".", "[.]", -1) + "$"}
+	}
+	svcPort := prule.Backend.ServicePort
+	svcName := prule.Backend.ServiceName
+
+	svc, err = state.getService(ns, svcName)
+	if err != nil {
+		log.Errorf("convertPathRule: Failed to get service %s, %s, %s", ns, svcName, svcPort)
+		return nil, err
+	}
+
+	targetPort, err := svc.getTargetPort(svcPort)
+	if err != nil {
+		// fallback to service, but service definition is wrong or no pods
+		log.Errorf("Failed to find target port for service %s, fallback to service: %v", svcName, err)
+		err = nil
+	} else if svc.Spec.Type == "ExternalName" {
+		scheme := "https"
+		if targetPort != "443" {
+			scheme = "http"
+		}
+		u := fmt.Sprintf("%s://%s:%s", scheme, svc.Spec.ExternalName, targetPort)
+		f, e := eskip.ParseFilters(fmt.Sprintf(`setRequestHeader("Host", "%s")`, svc.Spec.ExternalName))
+		if e != nil {
+			return nil, e
+		}
+		return &eskip.Route{
+			Id:          routeID(ns, name, "", "", svc.Spec.ExternalName),
+			BackendType: eskip.NetworkBackend,
+			Backend:     u,
+			Filters:     f,
+		}, nil
+	} else {
+		// err handled below
+		eps, err = state.getEndpoints(ns, svcName, svcPort.String(), targetPort)
+		log.Debugf("convertPathRule: Found %d endpoints %s for %s", len(eps), targetPort, svcName)
+	}
+
+	// TODO: errors may get ignored if err != nil && err != errEndpointNotFound && len(eps) == 0
+	if len(eps) == 0 || err == errEndpointNotFound {
+
+		address, err2 := getServiceURL(svc, svcPort)
+		if err2 != nil {
+			return nil, err2
+		}
+		r := &eskip.Route{
+			Id:          routeID(ns, name, host, prule.Path, svcName),
+			Backend:     address,
+			HostRegexps: hostRegexp,
+		}
+
+		setPath(pathMode, r, prule.Path)
+		setTraffic(r, svcName, prule.Backend.Traffic, prule.Backend.noopCount)
+		return r, nil
+
+	} else if err != nil {
+		return nil, err
+	}
+	log.Debugf("%d routes for %s/%s/%s", len(eps), ns, svcName, svcPort)
+
+	if len(eps) == 0 {
+		return nil, nil
+	}
+
+	// Consider: if there is only a single endpoint, wouldn't it be better to use the cluster IP?
```

Comment:
no because of DNAT/SNAT it's not safe for us to use the service IP

Reply:
got it.